### ISSUE TITLE
Issue with Notification binding after completing pay by link payments 

### ISSUE THREAD

**Issue by ankushKadlagMunro** [Adyen/adyen-dotnet-api-library#370]
HI Team,
We are trying to bind notification response to below
//This is our method (.Net framework)
[HttpPost]
public string
pdatePayByLinkStatus(Adyen.Model.Notification.NotificationRequest request)
{
}

We are referring your GitHub sample for this
https://github.com/Adyen/adyen-dotnet-api-library/blob/93a8c25608faeeddaf727dc941436c49152d51a5/Adyen/Model/Notification/NotificationRequest.cs#L30
and
https://docs.adyen.com/development-resources/webhooks/understand-notifications

After checking both the things we are not able to bind those notifications properly as class mentioned in dll has different properties than response received from Adyen backend.

if we directly use above classes it doesn't bind anything into "NotificationItemContainers"
and we get response below
{
Live: false
NotificationItemContainers:
}
so in short only first parameter is binding.
After this we tried to change second property name to "NotificationItems"
so its started binding something into it . After this we started getting below exception
We are getting below exception after we change properties name as per responses.

System.MissingMethodException: No parameterless constructor defined for this object. Object type 'Adyen.Model.Amount'

Can you please check and let us know ASAP.

**Comment by AlexandrosMor** [comment]
Hello @ankushKadlagMunro,

There is the notificationItems property which contains the list of NotificationRequestItem. Then the class that is containing all the rest of the properties is [NotificationRequestItem](https://github.com/Adyen/adyen-dotnet-api-library/blob/develop/Adyen/Model/Notification/NotificationRequestItem.cs). In the [unit test](https://github.com/Adyen/adyen-dotnet-api-library/blob/develop/Adyen.Test/NotificationTest.cs) the notification json is parsed properly to the classes.  
If this is not the class that you are missing could you please send us how you try to parse the notification ?
 
Kind regards,
Alexandros

**Comment by ankushKadlagMunro** [comment]
HI Alexandros,
Please see below.
Its still not working. We are trying by binding notification to our method like below (C# code)
[HttpPost]
public string UpdateLinkStatus(Adyen.Model.Notification.NotificationRequest request)
        {

}|

alternatively we have also tried by taking response in object in our method and converting it to NotificationRequest
JsonOperation.Deserialize<Adyen.Model.Notification.NotificationRequest>(request.ToString())

But you check below link from Adyen how they send us notification.
https://docs.adyen.com/development-resources/webhooks/understand-notifications
Here we are using Json method . we are not using soap.

if we bind them  its not binding any response to property “NotificationItemContainers” . It only bind first property.
 we get response below
{
Live: false
NotificationItemContainers:
}

Due to all this we are not able to Hmac verification of responses we get from adyen.
Can you please take it on high prio.



Thanks &  regards,
Ankush Kadlag
SelfMade, Software Private Limited
505 + 506 , Lunkad Sky Vista, New Airport Road, Viman Nagar,
Pune – 411014
Maharashtra, India
+91 20 69333406
[e-mail-logo_003]

Munro Tailoring
Koningin Wilhelminaplein 12 - C08
1062 HK Amsterdam
The Netherlands
+31 [0]20 4 085 008
www.munro-tailoring.com<http://www.munro-tailoring.com>

**Comment by AlexandrosMor** [comment]
Hello @ankushKadlagMunro,

Is the structure of the request that you are parsing similar to the notification request in [this unit test?](https://github.com/Adyen/adyen-dotnet-api-library/blob/develop/Adyen.Test/Mocks/notification/authorisation-true.json). Could you try to use the library to parse the notification similar to [here](https://github.com/Adyen/adyen-dotnet-api-library/blob/develop/Adyen.Test/NotificationTest.cs#L39) ?
We would like to investigate the issue further. Could you please send us the request that is not deserializing with `JsonOperation.Deserialize<Adyen.Model.Notification.NotificationRequest>(request.ToString())` so we can find out the iissue? 

regards,
Alexandros

**Comment by ankushKadlagMunro** [comment]
Hi Alexandros,
Below  is the example response which is not deserializing.
{"live":"false","notificationItems":[{"NotificationRequestItem":{"additionalData":{"expiryDate":"12\/2012"," NAME1 ":"VALUE1","authCode":"1234","cardSummary":"7777","totalFraudScore":"10","hmacSignature":"1XL2yi7UXnRQdZn\/0Pf1evVOEXnXBWsrqo7BK70soPg=","NAME2":"  VALUE2  ","fraudCheck-6-ShopperIpUsage":"10"},"amount":{"currency":"EUR","value":10100},"eventCode":"AUTHORISATION","eventDate":"2021-01-28T11:35:17+01:00","merchantAccountCode":"AMCustommadeClothingInc","merchantReference":"8313842560770001","operations":["CANCEL","CAPTURE","REFUND"],"paymentMethod":"visa","pspReference":"test_AUTHORISATION_1","reason":"1234:7777:12\/2012","success":"true"}}]}



Thanks &  regards,
Ankush Kadlag
SelfMade, Software Private Limited
505 + 506 , Lunkad Sky Vista, New Airport Road, Viman Nagar,
Pune – 411014
Maharashtra, India
+91 20 69333406
[e-mail-logo_003]

Munro Tailoring
Koningin Wilhelminaplein 12 - C08
1062 HK Amsterdam
The Netherlands
+31 [0]20 4 085 008
www.munro-tailoring.com<http://www.munro-tailoring.com>

**Comment by AlexandrosMor** [comment]
Hello @ankushKadlagMunro,
 
Could you please use those lines and check the notificationRequestItemContainer object ? 
`
 var handleNotificationRequest = notificationHandler.HandleNotificationRequest(jsonRequest);
 var notificationRequestItemContainer = handleNotificationRequest.NotificationItemContainers.FirstOrDefault();
  `
kind regards, 
Alexandros

**Comment by ankushKadlagMunro** [comment]
Hi Alexandros,
Below is our method which going to capture notifications from adyen system.
When using this method we are getting null in request object.

Ideally there should be a json string and upon deserialize it should convert into desired object.

We have tried different combinations but nothing is working till now.

Contacted Adyen system and they have redirected us to you.

Can you please help us to fix this issue asap?

[HttpPost]
        public string UpdateLinkStatus(string request)
        {
            Logger.Info("New UpdatePayByLinkStatus");
            if (request == null)
            {
                Logger.Info("null");

            }
            var notificationHandler = new Adyen.Notification.NotificationHandler();
            var handleNotificationRequest = notificationHandler.HandleNotificationRequest(request.ToString());
            var notificationRequestItemContainer = handleNotificationRequest.NotificationItemContainers.FirstOrDefault();
            return "accepted";
        }
we have below setting in Adyen backend for notification configurations. As already mentioned to you in earlier email we are using json way to receive response in our system.
but if you take request type as string it binds null when we receive response. can you check.
[cid:image002.png@01D6F630.CD3C1BE0]



Thanks &  regards,
Ankush Kadlag
SelfMade, Software Private Limited
505 + 506 , Lunkad Sky Vista, New Airport Road, Viman Nagar,
Pune – 411014
Maharashtra, India
+91 20 69333406
[e-mail-logo_003]

Munro Tailoring
Koningin Wilhelminaplein 12 - C08
1062 HK Amsterdam
The Netherlands
+31 [0]20 4 085 008
www.munro-tailoring.com<http://www.munro-tailoring.com>

**Comment by AlexandrosMor** [comment]
Hello @ankushKadlagMunro,

It is strange as for this part of code there is unit test which pass and deserialize the same json notification properly. Before moving further to investigate the issue, could you please confirm that you are receiving the notification from Adyen backend properly and that the string request of `UpdateLinkStatus` is not null ?

regards,
Alexandros

**Comment by ankushKadlagMunro** [comment]
HI Alexandros,
If we keep it as string then Adyen system shows us error like  object is moved and not binding it.  Its not able to find method.
but according to Adyen they sent us Json.

[cid:image001.png@01D6F63D.C3F04630]
. but if we use  object type  or our own class type  by converting above Json to class then it binds the result.

Please let me know if you need any more info.


Thanks &  regards,
Ankush Kadlag
SelfMade, Software Private Limited
505 + 506 , Lunkad Sky Vista, New Airport Road, Viman Nagar,
Pune – 411014
Maharashtra, India
+91 20 69333406
[e-mail-logo_003]

Munro Tailoring
Koningin Wilhelminaplein 12 - C08
1062 HK Amsterdam
The Netherlands
+31 [0]20 4 085 008
www.munro-tailoring.com<http://www.munro-tailoring.com>

**Comment by AlexandrosMor** [comment]
Hello @ankushKadlagMunro,

You do receive a json string from Adyen, right ?

**Comment by ankushKadlagMunro** [comment]
Hi Alexandros,
Please see below .
Our code
[HttpPost]
        public string UpdateLinkStatus(string request)
        {
            Logger.Info("New UpdatePayByLinkStatus");
            if (request == null)
            {
                Logger.Info("null");

            }
            else
            {
                Logger.Info("not null ");
            }
            var notificationHandler = new Adyen.Notification.NotificationHandler();
            var handleNotificationRequest = notificationHandler.HandleNotificationRequest(request.ToString());
            var notificationRequestItemContainer = handleNotificationRequest.NotificationItemContainers.FirstOrDefault();
}

Now see below logs from our system and adyen as below
we  get null  in our system . see below logs are printed
[cid:image001.png@01D6F649.5A590FC0]

and Adyen system response on their backend

[cid:image002.png@01D6F649.5A590FC0]
but they sent json as below
[cid:image003.png@01D6F649.5A590FC0]
So in short its not binding to string .
and settings in adyen backend.
[cid:image004.png@01D6F649.5A590FC0]
 Please let me know if you need any further info
Thanks &  regards,
Ankush Kadlag
SelfMade, Software Private Limited
505 + 506 , Lunkad Sky Vista, New Airport Road, Viman Nagar,
Pune – 411014
Maharashtra, India
+91 20 69333406
[e-mail-logo_003]

Munro Tailoring
Koningin Wilhelminaplein 12 - C08
1062 HK Amsterdam
The Netherlands
+31 [0]20 4 085 008
www.munro-tailoring.com<http://www.munro-tailoring.com>

**Comment by AlexandrosMor** [comment]
Hello,

The images that you provide are not visible. Indeed if you try to parse the request as null it is not going to work. The issue here is that the web hook is not receiving notifications and this is not related to the library as it is only parsing the notification request when is present. I would recommend to test the webhook if it is accessible. 

regards,
Alexandros

**Comment by ankushKadlagMunro** [comment]
Hi Alexandros,
Please see attached.
what setting is needed in webhook to get response as string because don’t see any option . its only JSON,SOAP, HTTP post
I hope you can help us out.




Thanks &  regards,
Ankush Kadlag
SelfMade, Software Private Limited
505 + 506 , Lunkad Sky Vista, New Airport Road, Viman Nagar,
Pune – 411014
Maharashtra, India
+91 20 69333406
[e-mail-logo_003]

Munro Tailoring
Koningin Wilhelminaplein 12 - C08
1062 HK Amsterdam
The Netherlands
+31 [0]20 4 085 008
www.munro-tailoring.com<http://www.munro-tailoring.com>

**Comment by AlexandrosMor** [comment]
Hello @ankushKadlagMunro 

The library can handle the notification after you received it but in your case you do not receive something. You need to investigate the issue from this point.
This issue is not caused by the library so for that reason I have to close the ticket.

kind regards,
Alexandros

**Comment by ankushKadlagMunro** [comment]
Hi Brayn,
I have escalated this issue to your plugin team and following since last week. (refer attached some  last thread communication)
According to them issue is not there with  library ,but issue is with notification we are getting from Adyen.
As per them we are not getting notification as string from your side so library is not able to convert it to json.

and please refer below code which we use to receive notification. Which we also send to GitHub. Please refer attached email for what happened when we write below code
Our code
[HttpPost]
        public string UpdateLinkStatus(string request)
        {
            Logger.Info("New UpdatePayByLinkStatus");
            if (request == null)
            {
                Logger.Info("null");

            }


Can you please  help us to solve this issue ASAP. As now this going from many days and we are not able to get solution to this.
This really causing delays. Can you please communicate to your plugin team and get us resolution to this

Please let us know.


Thanks &  regards,
Ankush Kadlag
SelfMade, Software Private Limited
505 + 506 , Lunkad Sky Vista, New Airport Road, Viman Nagar,
Pune – 411014
Maharashtra, India
+91 20 69333406
[e-mail-logo_003]

Munro Tailoring
Koningin Wilhelminaplein 12 - C08
1062 HK Amsterdam
The Netherlands
+31 [0]20 4 085 008
www.munro-tailoring.com<http://www.munro-tailoring.com>

From: Munro - Ankush
Sent: Tuesday, January 12, 2021 16:40
To: Adyen Support <support@adyen.com>
Subject: RE: [Adyen] Re: Notification binding is not working correctly

HI Bryan,
Thanks for reply
It would be nice if you can also escalate to your plugin team and get response .
This will help us in further process.




Thanks &  regards,
Ankush Kadlag
SelfMade, Software Private Limited
505 + 506 , Lunkad Sky Vista, New Airport Road, Viman Nagar,
Pune – 411014
Maharashtra, India
+91 20 69333406
[e-mail-logo_003]

Munro Tailoring
Koningin Wilhelminaplein 12 - C08
1062 HK Amsterdam
The Netherlands
+31 [0]20 4 085 008
www.munro-tailoring.com<http://www.munro-tailoring.com>

From: Adyen Support <support@adyen.com<mailto:support@adyen.com>>
Sent: Monday, January 11, 2021 07:59
To: Munro - Ankush <Ankush@munro-tailoring.com<mailto:Ankush@munro-tailoring.com>>
Subject: [Adyen] Re: Notification binding is not working correctly

------##- Please type your reply above this line -##------
[Adyen]

Hi Ankush,

Your request (#1797059<https://www.adyen.help/hc/requests/1797059>) has been updated. To add additional comments, reply to this email.
[https://secure.gravatar.com/avatar/0200eda0575e52c8a073817938d88d6c?size=40&default=https%3A%2F%2Fassets.zendesk.com%2Fimages%2F2016%2Fdefault-avatar-80.png&r=g]

Bryan Cresswell (Adyen)

Jan 11, 2021, 3:28 GMT+1
Hey Ankush,

As this is a library related issue, can you please raise an issue up on our .NET Github page here: https://github.com/Adyen/adyen-dotnet-api-library/issues

The developers will pick it up much quicker.

Let me know if you have additional questions.

Best Regards,

Bryan Cresswell
Technical Support Engineer

How would you rate this response?<https://www.adyen.help/requests/1797059/satisfaction/new/K0GoxlByXHTnUQKLubDzujooW?locale=1>
This email is a service from Adyen.
[ERKED8-5DQR]

**Comment by ankushKadlagMunro** [comment]
Hi Carm,
We  are facing this issue for “MunroTailoring099” on test env. Then we will check for Production.
Please let me know if you need anymore info




Thanks &  regards,
Ankush Kadlag
SelfMade, Software Private Limited
505 + 506 , Lunkad Sky Vista, New Airport Road, Viman Nagar,
Pune – 411014
Maharashtra, India
+91 20 69333406
[e-mail-logo_003]

Munro Tailoring
Koningin Wilhelminaplein 12 - C08
1062 HK Amsterdam
The Netherlands
+31 [0]20 4 085 008
www.munro-tailoring.com<http://www.munro-tailoring.com>

From: Adyen Support <support@adyen.com>
Sent: Wednesday, February 3, 2021 23:18
To: Munro - Ankush <Ankush@munro-tailoring.com>
Cc: Munro - Nipam <nipam@munro-tailoring.com>; Adyen/adyen-dotnet-api-library <reply@reply.github.com>; Adyen/adyen-dotnet-api-library <adyen-dotnet-api-library@noreply.github.com>; Mention <mention@noreply.github.com>
Subject: [Adyen] Re: RE: [Adyen] Re: Notification binding is not working correctly

------##- Please type your reply above this line -##------
[Adyen]

Hi Ankush,

Your request (#1854882<https://www.adyen.help/hc/requests/1854882>) has been updated. To add additional comments, reply to this email.
[https://secure.gravatar.com/avatar/fb75dd74f159b8e1f588f2890b6a1373?size=40&default=https%3A%2F%2Fassets.zendesk.com%2Fimages%2F2016%2Fdefault-avatar-80.png&r=g]

Carm DeMaio (Adyen)

Feb 3, 2021, 18:48 GMT+1
Hi Ankush,

What Merchant Account is experiencing this issue and on which environment? I'll take a look on our end to see what is going on.

Best regards,
________________________________

Carm DeMaio
Technical Support Engineer

Adyen San Francisco
274 Brannan Street Suite 600
San Francisco, CA 94107, United States

How would you rate this response?<https://www.adyen.help/requests/1854882/satisfaction/new/yYSeoxXTkVocJRMFpiK4hIcW8?locale=1>
This email is a service from Adyen.
[GD0V33-G56Q]

**Comment by bryancresswell** [comment]
Hi Ankush, can you please send an e-mail to support@adyen.com? This page is only meant for issues related to development.
